### PR TITLE
refactor the GP framework 4

### PR DESCRIPTION
**Pull request by yorkerlin**

### CODE REVIEW

File: src/interfaces/modular/Minimizer_includes.i [hunk 1]
```diff
@@ -0,0 +1,5 @@
+%{
+ #include <shogun/optimization/Minimizer.h>
+ #include <shogun/optimization/FirstOrderMinimizer.h>
+ #include <shogun/optimization/lbfgs/LBFGSMinimizer.h>
```
Comment (yorkerlin) [inline src/interfaces/modular/Minimizer_includes.i]: @karlnapf 
If I want to add nlopt in the i file, how can I guard nlopt when `nlopt` is missing/disabled?

File: src/interfaces/modular/Minimizer.i [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+
+/* Remove C Prefix */
+
+/* These functions return new Objects */
+
+/* Include Class Headers to make them visible from within the target language */
+%include <shogun/optimization/Minimizer.h>
+%include <shogun/optimization/FirstOrderMinimizer.h>
+%include <shogun/optimization/lbfgs/LBFGSMinimizer.h>
```
Comment (yorkerlin) [inline src/interfaces/modular/Minimizer.i]: @karlnapf
If I want to add nlopt in the i file, how can I guard nlopt when nlopt is missing/disabled?

File: src/shogun/machine/gp/SingleLaplacianInferenceMethod.cpp [hunk 3]
```diff
@@ -295,97 +492,54 @@ void CSingleLaplacianInferenceMethod::update_init()
...
+	SingleLaplacianNewtonOptimizer *opt=dynamic_cast<SingleLaplacianNewtonOptimizer*>(m_minimizer);
+	if (opt)
+	{
+		SG_REF(this);
+		opt->set_target(this);
+		opt->minimize();
+	}
+	else
+	{
+		FirstOrderMinimizer* minimizer= dynamic_cast<FirstOrderMinimizer*>(m_minimizer);
```
Comment (yorkerlin) [inline src/shogun/machine/gp/SingleLaplacianInferenceMethod.cpp]: @karlnapf 
Now, lbfgs and nlopt minimizers can be used in `SingleLaplacianInferenceMethod`.

File: src/shogun/machine/gp/SingleLaplacianInferenceMethod.cpp [hunk 4]
```diff
@@ -295,97 +492,54 @@ void CSingleLaplacianInferenceMethod::update_init()
...
+        if(minimizer!=m_minimizer)
+        {
+                delete m_minimizer;
+                m_minimizer=minimizer;
+        }
+}
 
+void CSingleLaplacianInferenceMethod::update_alpha()
+{
+	SingleLaplacianNewtonOptimizer *opt=dynamic_cast<SingleLaplacianNewtonOptimizer*>(m_minimizer);
```
Comment (yorkerlin) [inline src/shogun/machine/gp/SingleLaplacianInferenceMethod.cpp]: The default Newton minimizer.

File: src/shogun/machine/gp/SingleLaplacianInferenceMethod.cpp (hunk 3 above)
Comment (karlnapf) [inline src/shogun/machine/gp/SingleLaplacianInferenceMethod.cpp]: This is very good!

**Pull request merged**